### PR TITLE
Improve transfer warning flow

### LINKED ISSUE #2620: Send page: no invalid address error
If an invalid address is entered, the address field changes its border color to red, but no error message is displayed:
image

An error message appears only when an amount is entered:
image

Also, what else could cause an "Incorrect transaction information" error? I can't imagine anything else other than an invalid address. Therefore, I suggest changing the message error to: "Destination address is not a valid Monero address"

### PR DESCRIPTION
**Pull request by tobtoht**
closes #2620

### CODE REVIEW

File: pages/Transfer.qml [hunk 1]
```diff
@@ -782,10 +776,19 @@ Rectangle {
...
+
+        // Address is valid
+        if(addressLine.text && !walletManager.addressValid(addressLine.text, appWindow.persistentSettings.nettype)){
+            root.sendButtonWarning = qsTr("Address is invalid.") + translationManager.emptyString;
+            return false;
+        }
+
+        // Amount is nonzero
+        if (!amountLine.text || parseFloat(amountLine.text) <= 0) {
+            root.sendButtonWarning = qsTr("Enter an amount.") + translationManager.emptyString;
```
Comment (rating89us) [inline pages/Transfer.qml]: This error message is never displayed because send button is disabled when amount is empty.

File: pages/Transfer.qml [hunk 2]
```diff
@@ -772,7 +766,7 @@ Rectangle {
         }
 
         // There are sufficient unlocked funds available
-        if(parseFloat(amountLine.text) > parseFloat(middlePanel.unlockedBalanceText)){
+        if(parseFloat(amountLine.text) > parseFloat(walletManager.displayAmount(currentWallet.balance(currentWallet.currentSubaddressAccount)))){
             root.sendButtonWarning = qsTr("Amount is more than unlocked balance.") + translationManager.emptyString;
```
Comment (rating89us) [inline pages/Transfer.qml]: This error message is also never displayed.

User only gets a pop up with 
`Can't create transaction: not enough money to transfer, overall balance only 0.000000000000, sent amount 1111111.000000000000`

**Comment by tobtoht** [comment]
Please see http://xmrguide42y34onq.onion/static/warnings.mp4

**Comment by rating89us** [comment]
Yeah, nevermind, I misread your code

**Comment by rating89us** [comment]
Amount field border should change to red when the amount error message is displayed.

**Comment by selsta** [comment]
@tobtoht Btw can add a prefix to your commit titles in the future? E.g. `Transfer: improve warning flow` Don’t change your current open PRs but it would be nice to do it in the future :)

**Comment by tobtoht** [comment]
@selsta Certainly.

File: pages/Transfer.qml [hunk 3]
```diff
@@ -782,10 +776,19 @@ Rectangle {
...
-        // The transactional information is correct
-        if(!pageRoot.checkInformation(amountLine.text, addressLine.text, paymentIdLine.text, appWindow.persistentSettings.nettype)){
-            if(amountLine.text && addressLine.text)
-                root.sendButtonWarning = qsTr("Transaction information is incorrect.") + translationManager.emptyString;
+        if (addressLine.text == "") {
+            return false;
+        }
+
+        // Address is valid
+        if(addressLine.text && !walletManager.addressValid(addressLine.text, appWindow.persistentSettings.nettype)){
```
Comment (xiphon) [inline pages/Transfer.qml]:
```suggestion
        if (!walletManager.addressValid(addressLine.text, appWindow.persistentSettings.nettype)) {
```

File: pages/Transfer.qml [hunk 4]
```diff
@@ -772,7 +766,7 @@ Rectangle {
         }
 
         // There are sufficient unlocked funds available
-        if(parseFloat(amountLine.text) > parseFloat(middlePanel.unlockedBalanceText)){
+        if(parseFloat(amountLine.text) > parseFloat(walletManager.displayAmount(currentWallet.balance(currentWallet.currentSubaddressAccount)))){
```
Comment (selsta) [inline pages/Transfer.qml]: `currentWallet` might not be defined yet when opening the wallet. Consider adding something like this:

```
diff --git a/pages/Transfer.qml b/pages/Transfer.qml
index 47ec9c4..42a5c15 100644
--- a/pages/Transfer.qml
+++ b/pages/Transfer.qml
@@ -758,6 +758,9 @@ Rectangle {
         // reset message
         root.sendButtonWarning = "";
 
+        if (!currentWallet)
+            return true;
+
         // Currently opened wallet is not view-only
         if(appWindow.viewOnly){
             root.sendButtonWarning = qsTr("Wallet is view-only and sends are not possible. Unless key images are imported, " + 
```

**Comment by selsta** [comment]
Setting the amount to `all` results in an error now.

**Comment by tobtoht** [comment]
Disregard this latest push, this needs some more work.

**Comment by tobtoht** [comment]
Some extra changes: 

* Fixes a typo in WalletManager:maximumAllowedAmountAsString()
* Simplified libwalletqt/Wallet.cpp balance API
* Balance on balance card and notice on transfer page are truncated to a minimum of 2 decimals (relevant to #2457)

balance1
balance2

Sorry for going a bit out of scope with this. I can create a separate PR if needed.

**Comment by rating89us** [comment]
> Amount field border should change to red when the amount error message is displayed.

Is this being done in your code? Sorry, but I can't test it right now.

**Comment by tobtoht** [comment]
@rating89us Yes.

File: src/libwalletqt/WalletManager.cpp [hunk 5]
```diff
@@ -270,6 +270,12 @@ QString WalletManager::displayAmount(quint64 amount) const
     return QString::fromStdString(Monero::Wallet::displayAmount(amount));
 }
 
+QString WalletManager::displayAmountTruncated(quint64 amount) const
```
Comment (xiphon) [inline src/libwalletqt/WalletManager.cpp]: Move to qml

**Comment by xiphon** [comment]
> Sorry for going a bit out of scope with this. I can create a separate PR if needed.

Yes. Please split all the unrelated changes into separate PRs

**Comment by tobtoht** [comment]
If you don't mind I'll create a separate PR for the amount truncation. If I untangle this further it will become a dependency mess.

File: src/libwalletqt/Wallet.cpp [hunk 6]
```diff
@@ -289,6 +289,11 @@ bool Wallet::viewOnly() const
     return m_walletImpl->watchOnly();
 }
 
+quint64 Wallet::balance() const
+{
+    return m_walletImpl->balance(m_currentSubaddressAccount);
```
Comment (xiphon) [inline src/libwalletqt/Wallet.cpp]:
```suggestion
    return balance(m_currentSubaddressAccount);
```

File: src/libwalletqt/Wallet.cpp [hunk 7]
```diff
@@ -299,6 +304,11 @@ quint64 Wallet::balanceAll() const
     return m_walletImpl->balanceAll();
 }
 
+quint64 Wallet::unlockedBalance() const
+{
+    return m_walletImpl->unlockedBalance(m_currentSubaddressAccount);
```
Comment (xiphon) [inline src/libwalletqt/Wallet.cpp]:
```suggestion
    return unlockedBalance(m_currentSubaddressAccount);
```

Comment (tobtoht) [reply src/libwalletqt/Wallet.cpp]: Thanks

File: js/TxUtils.js [hunk 8]
```diff
@@ -40,6 +40,13 @@ function check256(str, length) {
     return true;
 }
 
+function getUnlockedBalance() {
```
Comment (xiphon) [inline js/TxUtils.js]: It is wallet-related thus has to be in main.qml

**Pull request merged**